boa/src/syntax/lexer/number.rs

```diff
@@ -138,84 +138,80 @@ impl<R> Tokenizer<R> for NumberLiteral {
         let c = cursor.peek();
 
         if self.init == '0' {
-            match c {
-                None => {
-                    // DecimalLiteral lexing.
-                    // Indicates that the number is just a single 0.
-                    return Ok(Token::new(
-                        TokenKind::NumericLiteral(Numeric::Integer(0)),
-                        Span::new(start_pos, cursor.pos()),
-                    ));
-                }
-                Some(Err(e)) => {
-                    todo!();
-                }
-                Some(Ok('x')) | Some(Ok('X')) => {
-                    // Remove the initial '0' from buffer.
-                    cursor.next();
-                    buf.pop();
-
-                    // HexIntegerLiteral
-                    kind = NumericKind::Integer(16);
-                }
-                Some(Ok('o')) | Some(Ok('O')) => {
-                    // Remove the initial '0' from buffer.
-                    cursor.next();
-                    buf.pop();
-
-                    // OctalIntegerLiteral
-                    kind = NumericKind::Integer(8);
-                }
-                Some(Ok('b')) | Some(Ok('B')) => {
-                    // Remove the initial '0' from buffer.
-                    cursor.next();
-                    buf.pop();
-
-                    // BinaryIntegerLiteral
-                    kind = NumericKind::Integer(2);
-                }
-                Some(Ok('n')) => {
-                    cursor.next();
-
-                    // DecimalBigIntegerLiteral '0n'
-                    return Ok(Token::new(
-                        TokenKind::NumericLiteral(Numeric::BigInt(0.into())),
-                        Span::new(start_pos, cursor.pos()),
-                    ));
-                }
-                Some(Ok(ch)) => {
-                    if ch.is_digit(8) {
-                        // LegacyOctalIntegerLiteral
-                        if self.strict_mode {
-                            // LegacyOctalIntegerLiteral is forbidden with strict mode true.
```

Comment:
@Razican 

I modified the lexer cursor peek to return  Option<Result<char, io::Error>> rather than  Option<&io::Result<char>>. 

This means that char is copied on return from peek however it provides the following:
 1. We don't have to store an error in peeked, if there is an io error it is returned immediately and can be dealt with (otherwise you jam peeked with an error value which should have been handled already and might now be fixed).
 
 2. We can remove areas where the result of peek is compared in a match branch using lots of Some(Ok()) such as the example here - this should hopefully provide a small performance boost which should offset any losses from copying char.